File: content/problems/2658-maximum-number-of-fish-in-a-grid/solutions/flood-fill-component-sums/Solution.java
class Solution {
    public int findMaxFish(int[][] grid){
        int rows=grid.length,cols=grid[0].length,answer=0;
        int[] directions={
            -1,0,1,0,-1
        };
        for(int startRow=0;startRow<rows;startRow++)for(int startCol=0;startCol<cols;startCol++)if(grid[startRow][startCol]>0){
            Queue<int[]> queue=new ArrayDeque<>();
            int fish=grid[startRow][startCol];
            grid[startRow][startCol]=0;
            queue.add(new int[]{
                startRow,startCol
            });
            while(!queue.isEmpty()){
                int[] cell=queue.poll();
                for(int d=0;d<4;d++){
                    int row=cell[0]+directions[d],col=cell[1]+directions[d+1];
                    if(row>=0&&row<rows&&col>=0&&col<cols&&grid[row][col]>0){
                        fish+=grid[row][col];
                        grid[row][col]=0;
                        queue.add(new int[]{
                            row,col
                        });
                    }
                }
            }
            answer=Math.max(answer,fish);
        }
        return answer;
    }
}

File: content/problems/2658-maximum-number-of-fish-in-a-grid/solutions/flood-fill-component-sums/Solution.kt
class Solution {
    fun findMaxFish(grid:Array<IntArray>):Int{
        val rows=grid.size;
        val cols=grid[0].size;
        val directions=intArrayOf(-1,0,1,0,-1);
        var answer=0;
        for(startRow in 0 until rows)for(startCol in 0 until cols)if(grid[startRow][startCol]>0){
            val queue=java.util.ArrayDeque<IntArray>();
            var fish=grid[startRow][startCol];
            grid[startRow][startCol]=0;
            queue.add(intArrayOf(startRow,startCol));
            while(queue.isNotEmpty()){
                val cell=queue.removeFirst();
                for(d in 0 until 4){
                    val row=cell[0]+directions[d];
                    val col=cell[1]+directions[d+1];
                    if(row in 0 until rows&&col in 0 until cols&&grid[row][col]>0){
                        fish+=grid[row][col];
                        grid[row][col]=0;
                        queue.add(intArrayOf(row,col))
                    }
                }
            };
            answer=maxOf(answer,fish)
        };
        return answer
    }
}
